components/builder-depot/src/server.rs

```diff
@@ -2307,6 +2264,393 @@ fn target_from_headers(user_agent_header: &UserAgent) -> result::Result<PackageT
+    depot: DepotUtil,
+    depot_cli: DepotClient,
+    oauth_token: Option<String>,
+    upstream_depot: &str,
+    ident: OriginPackageIdent,
+    channel: Option<String>,
+) -> Result<OriginPackage> {
+    let token_ref = oauth_token.as_ref().map(|x| &**x);
+    let parent_path = depot.archive_parent(&ident);
+
+    match fs::create_dir_all(parent_path.clone()) {
+        Ok(_) => {}
+        Err(e) => {
+            error!("Unable to create archive directory, err={:?}", e);
+            return Err(Error::IO(e));
+        }
+    };
+
+    // This is pretty frustrating, but I'll explain why this seemingly useless struct
+    // is here. We want to pass None as the last parameter to fetch_package, because
+    // this is running on a server and there's no TTY, so we have no need for
+    // a progress bar. Passing None by itself doesn't work, because the compiler
+    // complains that it can't infer the type for None. Passing None::<DisplayProgress>
+    // doesn't work because the compiler complains that it doesn't know the size of
+    // that parameter. So in order to get this to compile, we need to create this 100%
+    // useless struct that implements the DisplayProgress trait, just so we can cast
+    // None to it when we call the function. If someone has a better way of doing this,
+    // fantastic. I will happily change this code and subscribe to your newsletter.
+    struct NoProgress {}
+    impl DisplayProgress for NoProgress {
+        fn size(&mut self, _size: u64) {}
+
+        fn finish(&mut self) {}
+    }
+
+    impl Write for NoProgress {
+        fn write(&mut self, _buf: &[u8]) -> io::Result<usize> {
+            Ok(0)
+        }
+
+        fn flush(&mut self) -> io::Result<()> {
+            Ok(())
+        }
+    }
+
+    match depot_cli.fetch_package(&ident, token_ref, &parent_path, None::<NoProgress>) {
+        Ok(mut archive) => {
+            let target_from_artifact = archive.target().map_err(Error::HabitatCore)?;
+            if !depot.config.targets.contains(&target_from_artifact) {
+                debug!(
+                    "Unsupported package platform or architecture {}.",
+                    &target_from_artifact
+                );
+                return Err(Error::UnsupportedPlatform(target_from_artifact.to_string()));
+            };
+
+            if depot.config.jobsrv_enabled {
+                if let Err(r) = check_circular_deps(&ident, &target_from_artifact, &mut archive) {
+                    return Err(Error::IronResponse(r));
+                }
+            }
+
+            let mut package_create = match OriginPackageCreate::from_archive(&mut archive) {
+                Ok(p) => p,
+                Err(e) => {
+                    info!("Error building package from archive: {:#?}", e);
+                    return Err(Error::HabitatCore(e));
+                }
+            };
+
+            let config = depot.config.clone();
+
+            if let Err(e) = process_upload_for_package_archive(
```

Comment:
It could be a lot less code change and cleaner if we use the package upload endpoint, instead of commonizing the code and calling it from here.   So eventually, we could take the upstream refresh module and run it from anywhere (eg, I could even see it moving to the originsrv instead of the depot after we do the S3 migration).  If we use the endpoint, we can leave the package upload and download code almost exactly as it is currently.

Reply:
I think this is a great idea, but it feels like it would require significant changes to this PR.  Perhaps that's something we can pick up in a future refactor.